File: application/app/src/main/java/com/yahyeet/boardbook/model/mock/repository/AbstractMockRepository.java
package com.yahyeet.boardbook.model.mock.repository;

import com.yahyeet.boardbook.model.entity.AbstractEntity;
import com.yahyeet.boardbook.model.repository.IRepository;
import com.yahyeet.boardbook.model.repository.IRepositoryListener;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Function;

/**
 * Abstract mock implementation of a repository
 *
 * @param <T>
 */
public class AbstractMockRepository<T extends AbstractEntity> implements IRepository<T> {
	List<T> mockDatabase = new ArrayList<>();
	private int idCount = 0;

	@Override
	public CompletableFuture<T> find(String id) {
		return CompletableFuture.supplyAsync(() -> {

			for (T t : mockDatabase) {
				if (t.getId().equals(id))
					return t;
			}

			throw new CompletionException(new Exception("No user found"));
		});
	}

	@Override
	public CompletableFuture<T> save(T entity) {
		Function<T, T> create = T -> {
			entity.setId(Integer.toString(idCount));
			idCount++;
			mockDatabase.add(entity);
			return entity;
		};

		return CompletableFuture.supplyAsync(() -> {
			if (entity.getId() == null) {
				return create.apply(entity);
			} else {
				AtomicBoolean found = new AtomicBoolean(false);
				mockDatabase.forEach(m -> {
					if (m.equals(entity)) {
						int index = mockDatabase.indexOf(m);
						mockDatabase.set(index, entity);
						found.set(true);
					}
				});

				if (!found.get()) {
					return create.apply(entity);
				}

				return entity;
			}
		});
	}

	@Override
	public CompletableFuture<Void> delete(String id) {
		return CompletableFuture.supplyAsync(() -> {
			for (int i = 0; i < mockDatabase.size(); i++) {
				if (id.equals(mockDatabase.get(i).getId())) {
					mockDatabase.remove(i);
					return null;
				}
			}

			throw new CompletionException(new Exception("Not a user in database"));
		});
	}

	@Override
	public CompletableFuture<List<T>> all() {
		return CompletableFuture.supplyAsync(() -> mockDatabase);
	}

	@Override
	public void addListener(IRepositoryListener listener) {
		// Stub
	}

	@Override
	public void removeListener(IRepositoryListener listener) {
		// Stub
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/mock/repository/AbstractMockRepository.kt
package com.yahyeet.boardbook.model.mock.repository

import com.yahyeet.boardbook.model.entity.AbstractEntity
import com.yahyeet.boardbook.model.repository.IRepository
import com.yahyeet.boardbook.model.repository.IRepositoryListener

import java.lang.reflect.Array
import java.util.ArrayList
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Function

open class AbstractMockRepository<T : AbstractEntity> : IRepository<T> {

    protected var mockDatabase: MutableList<T> = ArrayList()
    protected var idCount = 0

    override fun find(id: String): CompletableFuture<T> {
        return CompletableFuture.supplyAsync {

            for (t in mockDatabase) {
                if (t.id == id)
                    return@CompletableFuture.supplyAsync t
            }

            throw CompletionException(Exception("No user found"))
        }
    }

    override fun save(entity: T): CompletableFuture<T> {
        val create = { T ->
            entity.id = Integer.toString(idCount)
            idCount++
            mockDatabase.add(entity)
            entity
        }

        return CompletableFuture.supplyAsync {
            if (entity.id == null) {
                return@CompletableFuture.supplyAsync create . apply entity
            } else {
                val found = AtomicBoolean(false)
                mockDatabase.forEach { m ->
                    if (m == entity) {
                        val index = mockDatabase.indexOf(m)
                        mockDatabase[index] = entity
                        found.set(true)
                    }
                }

                if (!found.get()) {
                    return@CompletableFuture.supplyAsync create . apply entity
                }

                return@CompletableFuture.supplyAsync entity
            }
        }
    }

    override fun delete(entity: T): CompletableFuture<Void> {
        return CompletableFuture.supplyAsync {
            for (i in mockDatabase.indices) {
                if (entity.id == mockDatabase[i].id) {
                    mockDatabase.removeAt(i)
                    return@CompletableFuture.supplyAsync null
                }
            }

            throw CompletionException(Exception("Not a user in database"))
        }
    }

    override fun all(): CompletableFuture<List<T>> {
        return CompletableFuture.supplyAsync { mockDatabase }
    }

    override fun addListener(listener: IRepositoryListener<*>) {

    }

    override fun removeListener(listener: IRepositoryListener<*>) {

    }
}
